in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+
+```ts
+
+// new
+export interface ArtifactAndInstance {
+  artifact: ContractArtifact;
+  instance: ContractInstanceWithAddress;
+}
+
+// modified
+export class TxExecutionRequest {
+  constructor(
+    // All these are the same:
+    public origin: AztecAddress,
+    public functionSelector: FunctionSelector,
+    public firstCallArgsHash: Fr,
+    public txContext: TxContext,
+    public argsOfCalls: PackedValues[],
+    public authWitnesses: AuthWitness[],
+    // Add:
+    /**
+     * Transient capsules needed for this execution.
+     */
+    public capsules: Fr[][],
+    /**
+     * Transient contracts needed for this execution.
+     */
+    public transientContracts: ArtifactAndInstance[],
+  ) {}
+  // ...
+}
+
+// new
+export interface InstanceDeploymentParams {
+  constructorName: string;
+  constructorArgs: any;
+  salt: Fr;
+  publicKeysHash: Fr;
+  deployer: AztecAddress;
+}
+
+// new
+export interface DeploymentOptions {
+  registerClass?: boolean;
+  publicDeploy?: boolean;
+}
+
+export interface UserFunctionCall {
+  contractInstance: ContractInstanceWithAddress;
+  functionName: string;
+  args: any;
+  deploymentOptions?: DeploymentOptions;
+  contractArtifact?: ContractArtifact;
+  functionAbi?: FunctionAbi;
+}
+
+// new
+export interface UserRequest {
+  calls: UserFunctionCall[];
+  gasSettings?: GasSettings;
+  paymentMethod?: FeePaymentMethod;
+  from?: AztecAddress;
+  simulatePublicFunctions?: boolean;
+  executionResult?: ExecutionResult; // the raw output of a simulation that can be proven
+  tx?: Tx; // a proven tx to send
+}
+
+export type TxExecutionRequestAdapter = (builder: TxExecutionRequestBuilder, userRequest: UserRequest) => Promise<void>;
+
+export interface TxExecutionRequestComponent {
+  adaptTxExecutionRequest: TxExecutionRequestAdapter;
+}
+
+export interface FeePaymentMethod extends TxExecutionRequestComponent {
+  getEquivalentAztBalance(): Promise<Fr>;
+}
+
+export interface SimulationOutput {
+  tx: Tx;
+  result: DecodedReturn | [];
+  request: UserRequest;
+  executionResult: ExecutionResult;
+  publicOutput: PublicSimulationOutput;
+  privateOutput: NestedProcessReturnValues;
+  error?: SimulationError;
+}
+
+export interface UserAPI {
+  getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest>;
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+
+// Swap `EntrypointInterface` for `TxExecutionRequestComponent`
+export interface AccountInterface extends AuthWitnessProvider, TxExecutionRequestComponent {
+  getCompleteAddress(): CompleteAddress;
+  getAddress(): AztecAddress;
+  getChainId(): Fr;
+  getVersion(): Fr;
+
+}
+
+// unchanged
+export interface AuthWitnessProvider {
```

Comment:
As mentioned earlier, I think it is pretty useful to have the Wallet and Account use different cases here, since one have access to information that it can then use to populate the call to the other. e.g., the wallet know chainId and version and just compute the messageHash that is provided to the account.

Reply:
So are you thinking something like:
```ts
class Account implements AccountInterface, AccountKeyRotationInterface {
  // ...
}

class Wallet implements UserAPI {
  constructor (public account: Account)
  // ...
}
```